refinery/tool_manager/utils.py

```diff
@@ -11,3 +19,175 @@ class AdminFieldPopulator(admin.ModelAdmin):
     def __init__(self, model, admin_site):
         super(AdminFieldPopulator, self).__init__(model, admin_site)
         self.list_display = [field.name for field in model._meta.fields]
+
+
+class WorkflowAnnotationValidationError(Exception):
+    """
+    Custom exception class that accepts an `error_message` upon instantiation
+    """
+    def __init__(self, error_message):
+        super(WorkflowAnnotationValidationError, self).__init__(error_message)
+
+
+class FileTypeValidationError(Exception):
+    """
+    Custom exception class that accepts a `filetype` and `error` upon
+    instantiation and builds a relevant error message
+    """
+    def __init__(self, filetype, error):
+        error_message = (
+            "Couldn't properly fetch FileType: {}.\n"
+            "Valid FileTypes are {}.\n"
+            " {}".format(
+                filetype,
+                [f.name for f in FileType.objects.all()],
+                error)
+        )
+
+        super(FileTypeValidationError, self).__init__(error_message)
+
+
+@transaction.atomic
+def create_tool_definition_from_workflow(workflow_dictionary):
+    """
+    :param workflow_dictionary: dict of data that represents a Galaxy workflow
+    """
+
+    tool_definition = ToolDefinitionFactory(
+        name=workflow_dictionary["name"],
+        description=workflow_dictionary["annotation"]["description"],
+        tool_type=workflow_dictionary["tool_type"],
+        file_relationship=create_file_relationship_nesting(
+            workflow_dictionary["annotation"])
+    )
+
+    for output_file in workflow_dictionary["annotation"]["output_files"]:
+        try:
+            filetype = FileType.objects.get(
+                name=output_file["filetype"]["name"])
+        except(FileType.DoesNotExist, FileType.MultipleObjectsReturned) as e:
+            raise FileTypeValidationError(output_file["filetype"]["name"], e)
+        else:
+            tool_definition.output_files.add(
+                OutputFileFactory(
+                    name=output_file["name"],
+                    description=output_file["description"],
+                    filetype=filetype
+                    )
+                )
+
+    # TODO: figure out Parameter/GalaxyParameter stuff
+
+
+@transaction.atomic
+def create_file_relationship_nesting(workflow_annotation,
+                                     file_relationships=None):
```

Comment:
Tidy up comments/ docstring and have `file_relationships` be a mutable empty list